### ISSUE TITLE
Use BER encoding for records rather than DER

### ISSUE THREAD

**Issue by salcock** [wanduow/openli#9]
BER is a less strict method than DER, which lets us make certain optimizations w.r.t. pre-encoding portions of the ETSI output.

Especially, BER allows us to encode lengths using any number of bytes rather than being forced to use the most efficient encoding -- we can always use 4 bytes to encode a sequence number, instead of having to encode seqno < 128 using 1 byte, etc. This means we can pre-construct the entire PS header with fixed-size spaces available for seqno, timestamp that get filled in at encoding time. At encode time, we can just copy the entire pre-encoded header into a buffer to send, then fill in the correct values for the missing fields, which should be way faster than the inside-out encoding that we do right now for DER.

Things to consider:
 * Can agencies handle BER-encoded records or do we need to still have DER available as an encoding option? Definitely check with NZP at least before we put too much time into this.
 * If DER is still required, how do we design an API that supports both encoding methods?
 * CC Payload sequences are a bit trickier -- every sequence within the Payload object will need its length set to account for the size of the saved payload. On the plus side, there is no variable-length intercept-specific content in the payload sequences, so the offsets for each of the length fields are likely to be in the same place.
 * PS Headers can be pre-encoded in the seqtracker thread and pushed to the encoder threads to be filled in.
 * IRI Payload probably won't have too much worth pre-encoding but hopefully we aren't generating these records at high rates anyway.

**Comment by primal100** [comment]
Is there a reason why DER was chosen? The standards specify BER only:

The transferred messages are encoded to be binary compatible with the Basic Encoding Rules (BER) as per
Recommendation ITU-T X.690 [12]. For more details see also ETSI TS 133 108 [9], clause B.1.

From the following document, section 5.3:
https://www.etsi.org/deliver/etsi_ts/102200_102299/10223201/03.20.01_60/ts_10223201v032001p.pdf

So I would think the Monitoring Centres would expect BER so should be easier for them. I've never seen DER being used anywhere for HI records but I suppose it's possible there are some New Zealand national requirements that I'm not aware of.

DER is mentioned in the standards, but only as part of a transition step for encrypting payloads (ANNEX G), so if you want to support that feature, keeping DER around might be a good idea. But the end result is still BER encoded, even in that case.

I've worked in LI for years and just found this project. It's very interesting. I need to learn C so I can understand the code and maybe contribute. :)

**Comment by salcock** [comment]
Hi,

Thanks for getting in touch -- it's great for me to get feedback and thoughts from other folks who have worked in this space before and have genuine hands-on experience with LI.

The main reason I chose DER to begin with was a practical one -- we did not (and still don't) have direct access to a genuine LEMF instance that we could test the OpenLI output against. Instead, we did internal testing with our self-written decoder and then occasional remote testing with the NZ police to confirm that their system could also decode what we were creating.

The remote testing is a massive painpoint -- it required lots of coordination and planning, and trying to diagnose and debug failures via a middle man was awkward at best (especially when they're trying to keep most of the details of the LEMF system that they are using secret from you).

So early on, I chose to use DER because there is only one valid encoding for any given input -- it was just a way for me to try and reduce the number of things that we could get wrong when doing our initial interoperability testing. Now that we know we are constructing the messages correctly, we can (in theory) now do testing with BER instead and know that any failures must be due to an error or incompatibility with our encoding rather than any other part of the system.
